warehouse/oidc/models/gitlab.py

```diff
@@ -0,0 +1,288 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any
+
+from sqlalchemy import ForeignKey, String, UniqueConstraint
+from sqlalchemy.dialects.postgresql import UUID
+from sqlalchemy.orm import Query, mapped_column
+from sqlalchemy.sql.expression import func, literal
+
+from warehouse.oidc.errors import InvalidPublisherError
+from warehouse.oidc.interfaces import SignedClaims
+from warehouse.oidc.models._core import (
+    CheckClaimCallable,
+    OIDCPublisher,
+    PendingOIDCPublisher,
+    check_claim_binary,
+)
+
+
+def _check_ci_config_ref_uri(ground_truth, signed_claim, all_signed_claims):
+    # We expect a string formatted as follows:
+    #   gitlab.com/OWNER/REPO//WORKFLOW_PATH/WORKFLOW_FILE.yml@REF
```

Comment:
Is `REPO//WORKFLOW_PATH` right here?

Reply:
Yup, from the [`ci_config_ref_uri`](https://docs.gitlab.com/ee/ci/secrets/id_token_authentication.html#token-payload) field docs:

> The ref path to the top-level pipeline definition, for example, gitlab.example.com/my-group/my-project//.gitlab-ci.yml@refs/heads/main. [Introduced](https://gitlab.com/gitlab-org/gitlab/-/issues/404722) in GitLab 16.2. This claim is null unless the pipeline definition is located in the same project.

Reply:
Gotcha. That's a bit weird, but I guess it is what it is 🙂 

(We should confirm with GitLab that the double-slash here won't change in the future.)

Reply:
I think it makes sense considering you can have nested groups/subgroups. So your repo might be located at `gitlab.com/group/subgroup1/subgroup2/repo`. Then, the double-slash is useful to disambiguate between the repo name and the folder. For example:
`gitlab.com/group/subgroup1/subgroup2/repo//folder/file.yml`. Without the double-slash, `repo` might be a folder and `subgroup2` the name of the repo.